src/hotspot/share/opto/regmask.hpp

```diff
@@ -64,28 +75,300 @@ class RegMask {
+
+  // RM_SIZE, but in number of machine words
   static const unsigned int _RM_SIZE     = LP64_ONLY(RM_SIZE >> 1) NOT_LP64(RM_SIZE);
+
+  // RM_SIZE_MIN, but in number of machine words
+  static const unsigned int _RM_SIZE_MIN =
+      LP64_ONLY(((RM_SIZE_MIN + 1) & ~1) >> 1) NOT_LP64(RM_SIZE_MIN);
+
+  // The last index (in machine words) of the (static) array of register mask
+  // bits
   static const unsigned int _RM_MAX      = _RM_SIZE - 1U;
 
+  // Compute a best-effort (statically known) upper bound for register mask
+  // size in 32-bit words. When extending/growing register masks, we should
+  // never grow past this size.
+  static const unsigned int RM_SIZE_MAX =
+      (((RM_SIZE_MIN << 5) +                // Slots for machine registers
+        (max_method_parameter_length * 2) + // Slots for incoming arguments
+        (max_method_parameter_length * 2) + // Slots for outgoing arguments
+        BoxLockNode_SLOT_LIMIT +            // Slots for locks
+        64                                  // Padding, reserved words, etc.
+        ) + 31) >> 5; // Number of bits -> number of 32-bit words
+
+  // RM_SIZE_MAX, but in number of machine words
+  static const unsigned int _RM_SIZE_MAX =
+      LP64_ONLY(((RM_SIZE_MAX + 1) & ~1) >> 1) NOT_LP64(RM_SIZE_MAX);
+
+  // Sanity check
+  STATIC_ASSERT(RM_SIZE <= RM_SIZE_MAX);
+
+  // Ensure that register masks cannot grow beyond the point at which
+  // OptoRegPair can no longer index the whole mask
+  STATIC_ASSERT(OptoRegPair::can_fit((RM_SIZE_MAX << 5) - 1));
+
   union {
-    // Array of Register Mask bits.  This array is large enough to cover
-    // all the machine registers and all parameters that need to be passed
-    // on the stack (stack registers) up to some interesting limit.  Methods
-    // that need more parameters will NOT be compiled.  On Intel, the limit
-    // is something like 90+ parameters.
+    // Array of Register Mask bits.  This array is large enough to cover all
+    // the machine registers and usually all parameters that need to be passed
+    // on the stack (stack registers) up to some interesting limit. On Intel,
+    // the limit is something like 90+ parameters.
+
+    // Viewed as an array of 32-bit words
     int       _RM_I[RM_SIZE];
+
+    // Viewed as an array of machine words
     uintptr_t _RM_UP[_RM_SIZE];
   };
 
-  // The low and high water marks represents the lowest and highest word
-  // that might contain set register mask bits, respectively. We guarantee
-  // that there are no bits in words outside this range, but any word at
-  // and between the two marks can still be 0.
+  // In rare situations (e.g., "more than 90+ parameters on Intel"), we need to
+  // extend the register mask with dynamically allocated memory. We keep the
+  // base statically allocated _RM_UP, and arena allocate the extended mask
+  // (RM_UP_EXT) separately. Another, perhaps more elegant, option would be to
```

Comment:
```suggestion
  // (_RM_UP_EXT) separately. Another, perhaps more elegant, option would be to
```
Underscore for consistency? Or does it reference something else?

Reply:
Yes, thanks (typo).